csrc/id_model/id_graph.h

```diff
@@ -0,0 +1,191 @@
+ public:
+  ValGraph() = default;
+
+  ValGraph(const ValGraph& other);
+  ValGraph(ValGraph&& other) = default;
+
+  ValGraph& operator=(const ValGraph& other);
+  ValGraph& operator=(ValGraph&& other) = default;
+
+  ValGraph(bool propagate_through_exprs)
+      : propagate_through_exprs_(propagate_through_exprs) {}
+
+  // Returns the disjoint val set.
+  const DisjointSets<Val*>& disjointValSets() const {
+    return disjoint_vals_;
+  }
+
+  DisjointSets<Val*>& disjointValSets() {
+    return disjoint_vals_;
+  }
+
+  // Returns the disjoint Expr set.
+  const DisjointSets<Expr*>& disjointExprSets() const {
+    return disjoint_exprs_;
+  }
+
+  DisjointSets<Expr*>& disjointExprSets() {
+    return disjoint_exprs_;
+  }
+
+  // Return if there's a group entry in the graph for this expr
+  bool hasGroup(Expr* expr) const;
+
+  // Return if there's a group entry in the graph for this val
+  bool hasGroup(Val* val) const;
+
+  // Convert expr to its exprGroup, assert that it exists.
+  const ExprGroup& toGroup(Expr* expr) const;
+
+  // Convert iter domain to its IdGroup, assert that it exists.
+  const ValGroup& toGroup(Val* val) const;
+
+  // Return output/input iter domain groups of provided expr
+  // Note that the same IdGroup can show up multiple times, so the
+  // output type cannot be VectorOfUniqueEntries
+  std::vector<ValGroup> outputGroups(const ExprGroup& expr) const;
+  std::vector<ValGroup> inputGroups(const ExprGroup& expr) const;
+
+  // Recursively traverses uses of the IdGroups in 'of' and returns all
+  // ExprGroups that have a use in their definition of provided of IdGroups.
+  ExprGroups allUsesOf(const ValGroups& of) const;
+
+  // Recursively traverses definitions of the IdGroups in 'of' and returns all
+  // ExprGroups used in this history of defining the 'of' IdGroups.
+  ExprGroups allDefinitionsOf(const ValGroups& of) const;
+
+  //! Returns the pointer to expressions associated with the
+  //! definitions of the provided IterDomain group in the provided
+  //! mapping mode (if it exists). Nullptr is returned otherwise.
+  //!
+  //! The returned pointer is to a vector of vector of expressions. The
+  //! inner vector is proven to be equivalent based on the provided mode. The
+  //! outer vector are expression groups that are not equivalent based on the
+  //! provided mode, but produce one of the IterDomains within the same disjoint
+  //! Iter Domain set based on the provided mode.
+  const ExprGroups* getDefinitions(const ValGroup& id_group) const;
+
+  //! Same as iterDomainGroupDefinitions but for uses instead of
+  //! definitions
+  const ExprGroups* getUses(const ValGroup& id_group) const;
+
+  bool hasDefinitions(const ValGroup& id_group) const;
+
+  bool hasUses(const ValGroup& id_group) const;
+
+  std::string toString() const;
+
+  // Checks if the expression is a trivial operation where an input is simply an
+  // output of the transformation. Returns the mapped iter domains if found.
```

Comment:
I don't follow this comment. What do you mean by "transformation" here? Does it refer to `expr` or things around it?